### PR TITLE
Fix issues when using `arc_select()` with `filter_geom` parameter (#61)

### PR DESCRIPTION
**Pull request by elipousson**
I'm unsure if this was a regression I introduced in the last PR or something else but the `filter_geom` parameter is not working with `arc_select()` as far as I can tell. The PR doesn't solve the issue but it hopefully helps. Updates include:

- Swapping no longer available `st_as_json()` function for updated `as_esri_geometry()` function
- Copy `match_spatial_rel()` from `{arcgisutils}` (not exported)

But the function still errors on `parse_esri_json()`. Presumably there is some issue that needs to be fixed in `{arcgisutils}` and this should stay open until that is resolved.

I ran into this issue while revising the README (I had a student try to install the package this evening and encountered some issues because `pak::pkg_install()` didn't install the dependences by default). The PR also include a couple fixes for documentation issues that I noticed in the `devtools::check()` results.

### CODE REVIEW

**Comment by JosiahParry** [comment]
Rather than copy a function because it's not exported, we can export it. We want to prevent duplication whereever possible.
Let's break this down into more distinct parts. 

Presently we have issues with spatial filtering on geometries. Let's fix that first and save documentation for a later time. There is a function `prepare_spatial_filter()` that is in arcgisutils as well. This is used in `delete_features()` 

https://github.com/R-ArcGIS/arcgisutils/blob/494e17003c3fa5ca08ab2b5b9365126f01b15701/dev/spatial-filter.R#L40 

My suggestion is that we remove `match_spatial_rel()` and `prepare_spatial_filter()` from arcgisutils and bring it into `arcgislayers`. This is because I do not think we will have a need for spatial filters when we do geocoding or routing (which will be in their own packages). 

Then we work on addressing `prepare_spatial_filter()` with associated tests. I'll create PRs to move the functions over. Then we can adapt the previous PR and this one into the `prepare_spatial_filter()` function. 

And, most importantly, create some tests.....

**Comment by JosiahParry** [comment]
This goes deep. So many tiny little pieces! 

I figured out why you were getting errors on `parse_esri_json()`. The way I was counting the number of results ignored any spatial filters. So additional requests were being generated with no body essentialy. The function cannot handle parsing (mostly) empty json like so: 

```json
{
    "objectIdFieldName": "OBJECTID",
    "uniqueIdField": {
        "name": "OBJECTID",
        "isSystemMaintained": true
    },
    "globalIdFieldName": "",
    "geometryProperties": {
        "shapeAreaFieldName": "Shape__Area",
        "shapeLengthFieldName": "Shape__Length",
        "units": "esriDecimalDegrees"
    },
    "features": []
}
```

This part has been addressed here: 
https://github.com/R-ArcGIS/arcgislayers/blob/7a359f5a0f8f69486017f86e4a06d4e9a5923264/R/arc-select.R#L298

**Comment by elipousson** [comment]
This definitely isn't the small and tidy PR you were looking for but I think the changes are all good and I migrated everything over from your spatial-filter branch. The only thing I didn't quite understand and ended up handling slightly differently was the CRS handling. Can you clarify why you wanted to move it out of `prepare_spatial_filter()`?

Let me know if you want me to add some tests to this PR or if I should back this up somewhere and start over with some smaller piecemeal changes.

File: R/utils-spatial-filter.R [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+#' - esriSpatialRelIntersects
+#' - esriSpatialRelContains
+#' - esriSpatialRelCrosses
+#' - esriSpatialRelOverlaps
+#' - esriSpatialRelTouches
+#' - esriSpatialRelWithin
+#'
+#' @export
+#' @rdname spatial_filter
+match_spatial_rel <- function(predicate, error_call = rlang::caller_env()) {
```
Comment (JosiahParry) [inline R/utils-spatial-filter.R:109]: I'm still not convinced that you need to have this as an argument lol but its not a hill I'll die on.

**Comment by JosiahParry** [comment]
Thanks for your help here. I took the liberty to push some minor updates to your fork so I can merge it. It looks like I'm a morning person ☕ and you a night owl 🦉.

Regarding CRS checking, the idea is that a function should do as little as possible. It should take only the inputs it really needs and do _only_ one thing. In the case of these packages the idea is that everything should end up being in a list that can be representable as json using `jsonify::to_json(x, unbox = TRUE)` which will be put into `req_body_form()`. If we require that we pass in a feature layer object to a function that prepares json for a spatial filter we cannot use that function elsewhere. By reducing the complexibility of the function we increase its portability.

Heres the note I had drafted for myself in my local branch:

```
  #' We want to do 1 thing for each function. Functions should be about creating
  #' list objects that turn into JSON. These objects are integrated into the requests
  #' prepare_spatial_filter() creates a list object that turns into the appropriate JSON
  #'  it works with only a single geometry object. Doing CRS matching and checking
  #'  should be done _outside_ of the function. Incorporating it into the function is
  #'  doing too much.
```
Thanks for your help here! I'm smashing the merge button. I'll try and get some tests written (finally) this weekend. But if you want to get a head start in a PR, here's what I've been drafting. 

I think snapshot tests are really important here same with a stable and public feature layer. I will on monday probably draft a test plan.

```r
  furl <- "https://services.arcgis.com/P3ePLMYs2RVChkJx/ArcGIS/rest/services/USA_Counties_Generalized_Boundaries/FeatureServer/0"
  devtools::load_all()
  x <- arc_open(furl)

  # no args
  expect_snapshot(arc_select(x))

  # has field
  expect_snapshot(
    arc_select(
      x,
      fields = c("STATE_NAME", "FIPS")
    )
  )

  # has where
  expect_snapshot(
    arc_select(
      x,
      where = "STATE_NAME = 'California'"
    )
  )

  # has field & where
  ca <- arc_select(
    x,
    fields = c("STATE_NAME", "FIPS"),
    where = "STATE_NAME = 'California'"
  )

  expect_snapshot(ca)

  # filtered geometry
  # sfc

  sfc <- ca$geometry
  arc_select(x, filter_geom = sfc, predicate = "intersects")

  # bbox
  bb <- sf::st_bbox(sfc)
  arc_select(x, filter_geom = bb, predicate = "intersects")
  
  # sfg
  mp <- sf::st_union(sfc[1:5])
  arc_select(x, filter_geom = mp, predicate = "intersects")
  arc_select(x, filter_geom = bb, predicate = "intersects")
```

File: R/utils-spatial-filter.R (hunk 1 above)
Comment (elipousson) [reply R/utils-spatial-filter.R:109]: It is basically habit at this point more than some well-reasoned argument. But, basically, I read the [rlang docs around this topic](https://rlang.r-lib.org/reference/topic-error-call.html) a couple years back and started adding `call` or `error_call` to any function that includes any input checks. 🤷‍♂️

**Pull request merged**